hash.c

```diff
@@ -2678,6 +2678,33 @@ rb_hash_values_at(int argc, VALUE *argv, VALUE hash)
     return result;
 }
 
+/*
+ *  call-seq:
+ *    hash.delete_at(*keys) -> new_array
+ *
+ *  Deletes entries at the specified keys, and returns a new \Array
+ *  containing values for the given +keys+:
+ *    h = {foo: 0, bar: 1, baz: 2}
+ *    h.delete_at(:baz, :foo) # => [2, 0]
+ *    h # => {bar: 1}
+ *
+ *  The {default values}[#class-Hash-label-Default+Values] are returned
+ *  for any keys that are not found:
+ *    h.delete_at(:hello, :bar) # => [nil, 1]
+*/
+VALUE
+rb_hash_delete_at(int argc, VALUE *argv, VALUE hash)
+{
+    VALUE result = rb_ary_new2(argc);
+    long i;
+
+    for (i=0; i<argc; i++) {
+	rb_ary_push(result, rb_hash_aref(hash, argv[i]));
+	rb_hash_delete(hash, argv[i]);
```

Comment:
@composerinteralia the tests covering the default and default proc are failing:

https://github.com/ruby/ruby/pull/4193/checks?check_run_id=1923986202#step:15:188

Is there an obvious way to achieve the same sequence of fallbacks in this C code?

Reply:
```c
VALUE
rb_hash_delete_at(int argc, VALUE *argv, VALUE hash)
{
    VALUE result = rb_ary_new2(argc);
    VALUE value;
    long i;

    for (i=0; i<argc; i++) {
      value = rb_hash_aref(hash, argv[i]);
      if (NIL_P(value) && rb_block_given_p()) {
        value = rb_yield(Qnil);
      }

	rb_ary_push(result, value);
	rb_hash_delete(hash, argv[i]);
    }
    return result;
}
```

I hardly know what I'm doing here, but I think this would get those tests passing. Passing a block to delete_at makes it less similar to values_at. Should they match?

Reply:
> Passing a block to delete_at makes it less similar to values_at. Should they match?

Hmm good question. Support for the block argument draws inspiration from the `Hash#delete` signature, but maybe it doesn't need to?